README
SEPP - Software Deployment System
---------------------------------
Installation instructions
=========================
by Tobias Oetiker <tobi@oetiker.ch> (2006-02-08.1)
by Roman Plessl <roman.plessl@oetiker.ch> (2008-07-28)

* create a SEPP install directory and give it to your local SEPP user. 

  All SEPP applications will be installed as that user. Note that you 
  MUST use the directory /usr/sepp and /usr/pack as long as you want to 
  share SEPP packages with other people. /usr/sepp and /usr/pack are
  the default pathes.

  Often software relies on absolute path-names.

  # mkdir /usr/sepp; cd /usr/sepp
  # adduser sepp
  # chown sepp /usr/sepp 
  # su - sepp

* unpack the SEPP base installation into the /usr/sepp directory. 
  (the % character shows that the following things should be done
  as the SEPP user)

  % cd /usr/sepp
  % wget http://oss.oetiker.ch/op-sepp/pub/sepp-1.5.3.tar.bz2 -O - | bunzip2 | tar xf -

* Add the symlinks to perl and rsync by just add symbolic links:

  % cd /usr/sepp
  % ln -s /usr/bin/rsync sbin/rsync
  % ln -s /usr/bin/perl sbin/perl

* the next step is to setup your package repository. If you have only one
  SEPP partition, then this is simple:

  # mkdir /usr/pack
  # chown sepp /usr/pack

  If you are planning on a larger deployment where each SEPP package could
  be on a different partition or even a different host, then you have to
  setup the automounter to dynamically build /usr/pack based on the content
  of /usr/sepp/conf/autosepp_master this is quite system dependent. On
  Solaris you do
  
  # echo +/usr/sepp/conf/autosepp_master >>/etc/auto_master
  # automount -v

  On other systems you may want to use a program mount or some other means
  since SEPP assumes that a new /usr/pack directory will become visible as
  soon as it is added to the automount map. Check out conf/sepp.pl for
  program mounts. It works on Solaris and Linux.

* If you went down the automount path, you have to create a local package
  store:

  # mkdir /usr/pack-store-a
  # chown sepp /usr/pack-store-a

* SEPP has to know some things about your system so that it can work
  properly. These settings are all stored in /usr/sepp/conf/sepp.conf

  % cd /usr/sepp/conf  

  % cp sepp.conf-dist sepp.conf
  
  or if you work on a simple standalone system (no automount)
 
  % cp sepp.conf-standalone sepp.conf

  Now edit the file with an editor of your choice and make any changes you
  see fit.

  % vi sepp.conf

  A second config file is read at run-time by all sepp packages, providing
  them with some basic information about your system.

  % cp sepprc.system-dist sepprc.system
  % vi sepprc.system

* Make sure your  SEPP.packlists exists where sepp.conf says it should. And touch
  yourself an autosepp_indirect file into existence.

  % touch /usr/sepp/SEPP.packlist
  % touch /usr/sepp/conf/autosepp_indirect

* Create a init script running at machine startup to create the 
  /tmp/SEPP.OS.DETECTOR file owned by root or the SEPP user. 

  This init script should be done for security reasons!

  % vi sepp_osdetector.init.sh
  % cp sepp_osdetector.init.sh /etc/init.d/sepp_osdetector.sh

  add to wished runlevels like

  % ln -s ../init.d/sepp_osdetector.sh /etc/rc2.d/S50sepp_osdetector.sh

  change owner and permissions of sepp_osdetector.sh

  % chown root:root /etc/init.d/sepp_osdetector.sh
  % chmod 644       /etc/init.d/sepp_osdetector.sh

* You can now either start building your own sepp packages following the
  instructions in html/seppdoc.pdf or you can use sepp-get to fetch some
  packages of our public package server. 

  sepp-get is available from http://oss.oetiker.ch/optools.

If you are lucky to work in an environment where other people are using SEPP
as well, you might want to talk to them about connecting your SEPP setup
with theirs using their packlist (this requires an automount setup).

 


CHANGES
2012-xx-xx
----------
* released 1.6.0

2012-07-03
----------
* released 1.5.3
* added upstart autofs helper scripts
* added OS Compats for Ubuntu till 12.04
* change log infrastructure from local4 to local3 (dealing
  with esxi logs which are hardcoded to local4)
* added partition in debug output where package lives
  (trac ticket #7)
* fixed pod documentation about --seppdebug and --seppenv  
  (trac ticket #9)
* fixed OSDectector.conf and example in OSDetector.pm. Remark
  to full config file (trac ticket #10). 
  Updated Pod Documentation.  

2009-10-14
----------
* released 1.5.2
* added skip partition in determination of the best matching
  partition with free space. Skipping will be done if a file
  SKIP_FREE_DISK exists.   
* added ExistENV routine in SEPP::SeppStart
  (patch by Andreas Romer <andreas.romer@espros.ch>)
* added Ubunty Hardy, Debian Lenny in SEPP::OSDetector


2008-12-02
----------
* added SuSE SLES in OSDetector config file

2008-08-05
----------
* added compatiblity mode for all SEPP arch names for seppadm expose

2008-07-28
----------
* released 1.5.1
* added Documentation about SEPP::OSDetector and init script
* SEPP::OSDetector: compatibility with Config::Grammar 1.10 and newer
* SEPP::OSDetector: global configuration of SEPP user is given from 
  seppadm to SEPP::OSDetector now
* SEPP::OSDetector: fixed owner ship detection of /tmp/SEPP.OS.DETECTOR
* seppadm: chdir to / when switching user since the original directory
  may not be readable (eg /root)

2007-07-12
----------
* released 1.5.0
* addded SEPP::OSDetector and code cleanup for OS detection
* added seppadm license-expose functionality for known sepp-get customers

2006-05-01
----------
* released 1.4.2
* sepp-get fix rsync error
* documentation for MacOS-X

2006-03-03
----------
* released 1.4.1
* fixed locking problem on seppadm remove
* sepp-get fixes

2006-02-08
----------
* released 1.4.0
* changed versioning scheme to a linux-like version instead of the timestamp
* changed format of sepp-get's index.txt file
* added support for platform-specific dependencies. For example:
  ncurses-5.5-mo : i686-debian-linux3.1, sun4u-sun-solaris2.8

2006-02-08
----------
* released 20060208
* Integrated sepp-get (former sepprsync)
* Improved Install Instructions
* Added sepp.conf sample for standalone system
* write out dependencies when exposing a package

2006-02-07
----------
* released 20060207
* Added Support for OS X (see README.OSX)
* Added support for web-publishing applications with license 'world' using the new
  seppadm expose command. This works together with the sepprsync application
  from http://isg.ee.ethz.ch/tools/sepprsync
* Improved Base Population for SEPP directores (INSTALL, INSTALL.lib, start.pl)
* New Authorize.pm module to implement a simple license checking system

---
2001-09-18
----------
* released 20000324pl16
* report removed packages only if '--showremoved' is specified (telling if a
  package is removed does require all packages to be accessed and thus makes
  the report much slower)

2001-08-10
----------
* released 20000324pl15
* really read ~/.sepprc in SeppStart.pm (didn't work until now)

2001-08-10 Matthias Braendli <braendli@iis.ee.ethz.ch>
----------
* speling
* fix error in regular expression for hostname
* {PACKDIR} in sepp.conf should really be {PACK}: replace {PACK}
  too and make sepp.conf more clear about it
* allow code after variables definitions in sepprc.system

2001-08-10 Peter Bircher <peter.bircher@inf.ethz.ch>
----------
* added 'perl interpreter' to sepp.conf: it will be used to generate the stubs
* mailto: links in HTML generated for packages
* AFS fixes

2001-07-03
----------
* really fix Binaries list in HTML pages

2001-06-14
----------
* use 'R' instead of 'I' in report for installed but removed packages

2001-06-12
----------
* fix Binaries list in HTML pages
* released 20000324pl14

2001-04-24
----------
* remove REMOVABLE if a package is installed again after having been removed
* add '/' suffix to "HREF=$pack.SEPP" in generated html

2001-04-05
----------
* set DISPLAY and XAUTHORITY in conf/.cshrc

2001-03-18
----------
* fix starting of programs with name that contains two '-'
* released 20000324pl13

2001-03-17
----------
* fixed seppadm prepare pathname/pack
* released 20000324pl12

2001-03-15 tobi
----------
* modified autosepp_indirect processing
  - localhost will always be first
  - if localserver is in the list it will stay to gether with localhost,
    rest has to go
  - else all entries stay

2000-09-13
----------
* fixed "undefined values" in webbuild
* released 20000324pl10

2000-08-07
----------
* added 'use strict'
* added folding markers
* released 20000324pl9

2000-05-24
----------
* delete automount entry for packages that cannot be installed because
  the platform doesn't match.
* released 20000324pl8

2000-05-11
----------
* --shellonly didn't only start the shell :-( An 'exit' was missing.
* released 20000324pl7

2000-05-10
----------
* --force will also install already existing manpages.
* released 20000324pl6

2000-05-09
----------
* readded <BASE TARGET=...> elements in the HTML output (App Index).
  Now links in app-page open in whole browser and not in frame.
* made App Index generation more tolerant to unaccessible packages (will skip
  then now, instead of aborting)
* released 20000324pl5

2000-05-02
----------
* readded check for already existing package when a seppadm prepare (without
  --shellonly) is made.
* released 20000324pl4

2000-04-29
----------
* some minor enhancements in App. Catalog
* CheckSEPPrc only called once.
* released 20000324pl3

2000-04-28
----------
* made "SEPP Application Catalog" HTML4 compliant
* added "Zoom" for each index in App. Catalog.
* added "sepp domain" to the title of the App. Catalog.
* use of tables instead of DIV and CSS for borders in app-pages of App.
  Catalog (Netscape Communicator didn't display the border correctly in some
  cases).

sbin/Authorize.pm
#!/usr/sepp/bin/perl-5.8.4 -w

package Authorize;

BEGIN {
    # these variables will be automagically set when the script
    # is installed. Remove what you don't need...
    $::ISGTC_MAGIC_LIBDIR = "/usr/isgtc/lib";
    $::ISGTC_MAGIC_SYSCONFDIR = "/usr/sepp/conf";
    unshift @INC, "$::ISGTC_MAGIC_LIBDIR/perl";
}

use lib("$::ISGTC_MAGIC_LIBDIR/perl", "perl");

use strict;
use vars qw(%opt);

use Getopt::Long 2.25 qw(:config posix_default);
use Pod::Usage 1.14;

use Sys::Hostname 1.1;
use POSIX;
use IO::Socket;
use Config::Grammar 1.0;
use NetAddr::IP 3.24;

my $RCS_VERSION = '$Id: Authorize.pm 157 2005-06-06 09:56:07Z aschults $';

$|=1;

### function prototypes ###

sub main();

sub is_authorized();
sub check();
sub notify();

sub message($);

sub mygetenv();

sub checkexpiry($);
sub checknetgroup($);
sub checkuser($);
sub checkgroup($);
sub checkwarn($);
sub checkhost($); 
sub checknetwork($);

sub readcfg($);

### global variables ###

my $hostname;
my $fqhostname;
my $ip;
my %netgroups;
my $username;
my %groups;
my $timestamp;

my $allowusage;
my $continueusage;
my $warn_before_expiry = 10;

### license files ###

my $global_cfgfile  = "$::ISGTC_MAGIC_SYSCONFDIR/authorize.conf";
my $local_cfgfile  = "./license.conf";
my $global_cfg;
my $local_cfg;

### program ###

if ($0 =~ m|Authorize.pl$|) {
    main;
}

### functions ###
sub mygetenv() {
    # find out hostname
    $hostname = hostname;

    # find out fully-qualified hostname
    $fqhostname = $hostname;
    if ($fqhostname =~ /^[^.]+$/ ) {
        $fqhostname = (gethostbyname("$fqhostname"))[0];
    }

    # find out IP
    $ip = inet_ntoa(inet_aton($hostname));

    # find out netgroups
    my $cmd = "/usr/bin/nismatch host=$fqhostname netgroup.org_dir";
    if ( -f "/var/nis/NIS_COLD_START") {
       open(NISMATCH, "$cmd |") or do {
           warn "WARNING: can't execute '$cmd': $!\n";
       };
       while(my $l = <NISMATCH>) {
           $l =~ /^(\S+)\s+$fqhostname\s*$/ or do {
                warn "WARNING: unexpected output from nismatch: $l";
                next;
           };
           $netgroups{$1}=1;
       }
       close(NISMATCH);
     }

    # find out username
    $username = getpwuid($>);

    # find out unix groups
    my @grouplist = split / /, $);
    foreach my $item (@grouplist){
         $item = getgrgid $item;
         $groups{$item}=1;
    }

    # find out unix timestamp 
    $timestamp = time;
 }

sub showmessage($){
    my $message = shift;
    my $messagefunction = "SEPP::Message::$^O";
    print $messagefunction;
    eval "require $messagefunction";
    my $msg = $messagefunction->new();
    $msg->showmessage($message);
}



# check expiry date
sub checkexpiry($){
    my $datestring = shift;
    if (not defined($datestring)){return 1; exit;}
    my ($compyear, $compmonth, $compday) = split(/-/,$datestring);

    $compmonth -= 1;
    $compyear -= 1900;

    my($comptimestamp) = mktime(0, 0, 0, $compday, $compmonth, $compyear);
    if ($comptimestamp  > $timestamp){
	return 1;
    } else {
	return 0;
    }
}

# check netgroup
sub checknetgroup($){
    my $compnetgrouplist = shift;
    if (not defined($compnetgrouplist)){return 1; exit;}
    my @compnetgroups = split(/, /,$compnetgrouplist);

    foreach (@compnetgroups) {
	if (defined($netgroups{$_})) {
	    return 1;
	} else {
	    return 0;
	}
    }
}

# check user
sub checkuser($){
    my $compuserlist = shift;
    if (not defined($compuserlist)){return 1; exit;}
    my @compusers = split(/, /,$compuserlist);

    foreach (@compusers) {
	if ($username eq $_) {
	    return 1;
	} else {
	    return 0;
	}
    }
}

# check groups
sub checkgroup($){
    my $compgrouplist = shift;
    if (not defined($compgrouplist)){return 1; exit;}
    my @compgroups = split(/, /,$compgrouplist);

    foreach (@compgroups) {
	if (defined($groups{$_})) {
	    return 1;
	} else {
	    return 0;
	}
    }
}

# check warn date
sub checkwarn($){
    my $datestring = shift;
    if (not defined($datestring)){return 1; exit;}
    my ($compyear, $compmonth, $compday) = split(/-/,$datestring);

    $compmonth -= 1;
    $compyear -= 1900;

    my($comptimestamp) = mktime(0, 0, 0, $compday, $compmonth, $compyear);
    if ($comptimestamp  > ($timestamp + $warn_before_expiry * 24 * 3600)){
        return 1;
    } else {
        return 0;
    }
}

# check host
sub checkhost($){
  my $comphostname = shift;
  if (not defined($comphostname)){return 1; exit;}
  if (($comphostname eq $hostname) || ($comphostname eq $fqhostname)) {
	return 1;
  } else {
	return 0;
  }
}

# check network
sub checknetwork($){
    my $networkstring = shift;
    if (not defined ($networkstring)){return 1;exit;}
    my ($net,$netmask) = split(/\//, $networkstring);

    my $netip = new NetAddr::IP "$ip";
    if ($netip->within(new NetAddr::IP "$net", "netmask")) {
       return 1;
   } else {
       return 0;
   }
}

# read configuaration file
sub readcfg($){
    my $RE_MODEL        = '(soft|hard|crypt)';
    my $RE_LICENSENAME  = '(\S+\s?)*';
    my $RE_USERNAME     = '\S+(,\s?\S+)*';
    my $RE_DATE         = '[0-9]{4}-[01][0-9]-[0-3][0-9]';
    my $RE_HOSTNAME     = '.*'; #'(\S+.)*\S+\s((\S+.)*\S+\s)*';
    my $RE_NET          = '\d+\.\d+\.\d+\.\d+/\d+\.\d+\.\d+\.\d+';      # 192.168.116.12/255.255.255.0
    my $RE_GROUPS       = '(\S+,\s?)*\S+';

    my $cfgfile = shift;
    my $e       = '=';
    my $parser  = Config::Grammar->new(
      {
      _sections => [ 'system', 'allow', 'continue', 'abort text', 'continue text' ],
      system => {
		 _doc => 'license model',
		 _vars => [ 'model' ],
		 model =>{
			  _doc => "definition of the licensing model",
			  _example => "soft",
			  _re => $RE_MODEL,
			  _re_error =>
			  'model must be one of soft, hard or crypt' },
		},
      allow =>{
	       _doc => 'allow starting the licenses',
	       _sections => [ "/$RE_LICENSENAME/" ],
	       "/$RE_LICENSENAME/" =>
	       {
		_doc    => "the name of the license or the license feature",
		_example => "testlicense-1.0a-rp",
		_vars   => [ 'user', 'expiry', 'host', 'net', 'group', 'netgroup','warn' ],
		user     => {
			     _doc => "an unique (unix) user name",
			     _example => "rplessl",
			     _re => $RE_USERNAME,
			     _re_error =>
			     'user must be an unix username'
			    },
		expiry => {
			   _doc => "an date yyyy-mm-dd",
			   _example => "2005-09-01",
			   _re => $RE_DATE,
			   _re_error =>
			   'expiry must be an date'
			  }, 
		warn => {
			 _doc => "an date yyyy-mm-dd",
			 _example => "2005-09-01",
			 _re => $RE_DATE,
			 _re_error =>
			 'warn must be an date'
			},
		host => {
			 _doc => "an dns hostname",
			 _example => "aeryn.ee.ethz.ch",
			 _re => $RE_HOSTNAME,
			 _re_error =>
			 'hostname must be an dns hostname with dots'
			},
		net => {
			_doc => "Default network address in IP notation added with netmask",
			_example => "10.12.33.1/255.255.255.0",
			_re => $RE_NET,
			_re_error =>
			'hosts must be given by network address and network mask'
			},
		group => {
			  _doc => "Unix groups separated with commas",
			  _example => "ifa, biwi",
			  _re => $RE_GROUPS,
			  _re_error =>
			  'groups must be a comma sparated list'
			 },
		netgroup => {
			     _doc => "nis netgroups separated with commas",
			     _example => "ifa_t, biwi_t",
			     _re => $RE_GROUPS,
			     _re_error =>
			     'groups must be a comma sparated list'
			    },
	       },
	      },
      continue =>{
		  _doc => 'allow continuing the licenses',
		  _sections => [ "/$RE_LICENSENAME/" ],
		  "/$RE_LICENSENAME/" =>
		  {
		   _doc    => "the name of the license or the license feature",
		   _example => "testlicense-1.0a-rp",
		   _vars   => [ 'user', 'expiry', 'host', 'net', 'group', 'netgroup', 'warn'],
		   user     => {
				_doc => "an unique (unix) user name",
				_example => "rplessl",
				_re => $RE_USERNAME,
				_re_error =>
				'user must be an unix username'
			       },
		   expiry => {
			      _doc => "an date yyyy-mm-dd",
			      _example => "2005-09-01",
			      _re => $RE_DATE,
			      _re_error =>
			      'expiry must be an date'
			     },
		   warn => {
			    _doc => "an date yyyy-mm-dd",
			    _example => "2005-09-01",
			    _re => $RE_DATE,
			    _re_error =>
			    'warn must be an date'
			   },
		   host     => {
			     _doc => "an dns hostname",
			     _example => "aeryn.ee.ethz.ch",
			     _re => $RE_HOSTNAME,
			     _re_error =>
			     'hostname must be an dns hostname with dots'
			    },
		   net => {
			    _doc => "Default network address in IP notation added with netmask",
			    _example => "10.12.33.1/255.255.255.0",
			    _re => $RE_NET,
			    _re_error =>
			    'hosts must be given by network address and network mask'
			   },
		   group => {
			     _doc => "Unix groups separated with commas",
			     _example => "ifa, biwi",
			     _re => $RE_GROUPS,
			     _re_error =>
			     'groups must be a comma sparated list'
			    },
		   netgroup => {
				_doc => "nis netgroups separated with commas",
				_example => "ifa_t, biwi_t",
				_re => $RE_GROUPS,
				_re_error =>
				'groups must be a comma sparated list'
			       },
		  },
		 },
      'abort text' => {
		       _doc => 'Abort Text',
		       _text => {},
		      },
      'continue text' => {
			  _doc => 'Abort Text',
			  _text => {},
			 }
     });

    if ($opt{poddoc}) {
        print $parser->makepod if $opt{poddoc};
        exit 0;
    }

    my $cfg = $parser->parse($cfgfile) or die "ERROR: $parser->{err}\n";
    return $cfg;
}


sub main(){
    # parse options
    %opt = ();
    GetOptions(
        \%opt,       'help|h',
        'man',       'version',
        'verbose|v', 'poddoc'
      )
      or pod2usage(2);
    if ($opt{help}) { pod2usage(1) }
    if ($opt{man}) { pod2usage(-exitstatus => 0, -verbose => 2) }
    if ($opt{version}) { print "$RCS_VERSION\n"; exit(0) }

    my $allow;
    $allow = is_authorized();
    notify();
    if ($allow) { return 1; } else {return 0;}
}

sub check(){
    my $allow;
    $allow = is_authorized();
    notify();
    if ($allow) { return 1; } else {return 0;}
}


sub is_authorized(){
    # get environment
    mygetenv();

    $local_cfgfile="$main::PackVar/licence.conf" if defined $main::PackVar;

    $local_cfg=undef;
    if( -e $local_cfgfile)
    {
      die "ERROR: local config file $local_cfgfile is not readable\n" unless -r $local_cfgfile;
      $local_cfg = readcfg($local_cfgfile);
    }
    
    $global_cfg=undef;
    if( -e $global_cfgfile)
    {
      # read configfile
      die "ERROR: global config file $global_cfgfile is not readable\n" unless -r $global_cfgfile;
      $global_cfg = readcfg($global_cfgfile);
    }

    # Use defaults if no local config available.
    $local_cfg=$global_cfg unless $local_cfg;

    # If no config file exists at all, don't do authorization.
    return 1 unless $local_cfg;
    
    ### Check each Rule
    foreach (keys %{$local_cfg->{'allow'}}) {
        my $expiry    = $local_cfg->{'allow'}->{$_}->{'expiry'}   ;
        my $netgroup  = $local_cfg->{'allow'}->{$_}->{'netgroup'} ;
        my $user      = $local_cfg->{'allow'}->{$_}->{'user'}     ;
        my $group     = $local_cfg->{'allow'}->{$_}->{'group'}    ;
        my $warn      = $local_cfg->{'allow'}->{$_}->{'warn'}     ;
        my $host      = $local_cfg->{'allow'}->{$_}->{'host'}     ;
        my $net       = $local_cfg->{'allow'}->{$_}->{'net'}      ;

        if ( checkexpiry($expiry) &&
             checknetgroup($netgroup) &&
             checkuser($user) &&
             checkgroup($group) &&
             checkwarn($warn) &&
             checkhost($host) &&
             checknetwork($net)){
	    $allowusage += 1;
	}
    }

    ### Check each Rule
    foreach (keys %{$local_cfg->{'continue'}}) {
        my $expiry    = $local_cfg->{'continue'}->{$_}->{'expiry'}   ;
        my $netgroup  = $local_cfg->{'continue'}->{$_}->{'netgroup'} ;
        my $user      = $local_cfg->{'continue'}->{$_}->{'user'}     ;
        my $group     = $local_cfg->{'continue'}->{$_}->{'group'}    ;
        my $warn      = $local_cfg->{'continue'}->{$_}->{'warn'}     ;
        my $host      = $local_cfg->{'continue'}->{$_}->{'host'}     ;
        my $net       = $local_cfg->{'continue'}->{$_}->{'net'}      ;

        if ( checkexpiry($expiry) && 
             checknetgroup($netgroup) &&
             checkuser($user) &&
             checkgroup($group) &&
             checkwarn($warn) &&
             checkhost($host) &&
             checknetwork($net)){
	    $continueusage += 1;
        }
    }
    if ( defined($allowusage) || defined($continueusage) ) { return 1; } else { return 0; }
}


sub notify(){
    if (not defined($allowusage)) {
	if (defined($continueusage) && $local_cfg->{'system'}->{'model'} eq 'soft'){
	    ($_ = $local_cfg->{'continue text'}->{'_text'}) =~ s/\n/ /;
	    if (not defined($_)){ ($_ = $global_cfg->{'continue text'}->{'_text'}) =~ s/\n/ /;	}
	    showmessage($_);
	} else {
	    ($_ = $local_cfg->{'abort text'}->{'_text'}) =~ s/\n/ /;
	    if (not defined($_)){ ($_ = $global_cfg->{'abort text'}->{'_text'}) =~ s/\n/ /; }
	    showmessage($_);
	}
    }
}

1

__END__

=pod

=head1 NAME

  SEPP::Authorize - An SEPP authorization extension.

=head1 SYNOPSIS

B<is_authorized()>

B<notify()>

B<check()>

=head1 DESCRIPTION

SEPP::Authorize is an SEPP extension which checks the
autorization and permissions before starting a
commercial software product.

B<is_authorized()> checks if a user/computer is authorized to
run an application in SEPP. A FALSE answer means that
the application start should not be granted.

B<notify()> informs a user about the rules by signalizing a
messagebox with the rules (sometimes a hint that this program
is allowed to evaluate).

B<check()> includes both functions is_autorized() and notify().

=head1 CONFIGFILE

=head2 Sections

There will be three sections in the config file:

=over

=item o system

select licensing model soft (only produces a warning, can continue 
to evaluate), hard (can't start if not licensed) or crypt (like hard, 
but the software is encrypted -- crypt is for future extensions).

=item o allow

cases which allows an usage of this license: the lines for each case 
are evaluated as an "AND" and each case as an "OR" for the section
allow.

=item o continue

cases which allows an continued usage of this license, an messagebox
will be displayed: the lines for each case are evaluated as an "AND" 
and each case as an "OR" for the section continue.

=back

=head2 Example Configfile

  *** system ***
  model = soft   # soft:  only produces a warning, can continue to evaluate
                 # hard:  can't start if not licensed
                 # crypt: like hard, but the software is encrypted

  *** allow ***
  + first case (e.g. some members of a group)
    user      = aschults, dws, luki
    netgroup  = biwi_t
    expiry    = 2006-12-31

  + second case (e.g. the owner of the license)
    user      = aschults
    expiry    = 2005-12-31

  + third case (e.g. default host for this license)
    host      = aeryn.ee.ethz.ch
    net       = 129.132.67.0/255.255.255.0
    expiry    = 2005-12-31
    user      = rplessl

  + forth case (e.g. a unix group)
    group     = biwi
    expiry    = 2005-12-31

  + fifth case (e.g. allow whole nis+ groups)
    netgroup  = biwi_t
    expiry    = 2006-12-31 
    warn      = 2006-10-01

  *** continue ***
  + first case (e.g. some persons of a group)
    user      = aschults, dws, luki
    netgroup  = biwi_t
    expiry    = 2006-12-31


  *** abort text ***
  The application you have started is not allowed to run on your machine.

  *** continue text ***
  The application You started is not allowed to run on your machine.
  You can continue and evaluate the application by pressing Continue.

=head2 Available Checks for the Configuration File

The available check routines for the configfile or listed below
or can be extracted by starting the I</usr/sepp/bin/perl
Authorize.pm --poddoc>.

=over

=item B<user>

an unique (unix) user name

Example: user = rplessl

=item B<expiry>

an date yyyy-mm-dd

Example: expiry = 2005-09-01

=item B<host>

an dns hostname

Example: host = aeryn.ee.ethz.ch

=item B<net>

Default network address in IP notation added with netmask

Example: net = 10.12.33.1/255.255.255.0

=item B<group>

Unix groups separated with commas

Example: group = ifa, biwi

=item B<netgroup>

nis netgroups separated with commas

Example: netgroup = ifa_t, biwi_t

=item B<warn>

an date yyyy-mm-dd

Example: warn = 2005-09-01

=back

=head1 BUGS

None known.

=head1 COPYRIGHT

Copyright (c) 2005 by ETH Zurich. All rights reserved.

=head1 LICENSE

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

=head1 AUTHOR

S<Roman Plessl E<lt>rplessl@ee.ethz.chE<gt>>

=head1 HISTORY

 2004-04-29 rp Initial Version

=cut

#
# Local Variables:
# mode: cperl
# eval: (cperl-set-style "PerlStyle")
# mode: flyspell
# mode: flyspell-prog
# End:
#
# vi: sw=4 et
